### PR TITLE
Take ownership of k8s-at-home Helm chart, 

### PR DESCRIPTION
**Pull request by alexander-bauer**
### Background

As discussed in #1790, folks are using Paperless on Kubernetes. Helm is a packager manager of sorts for Kubernetes, and a number of Helm charts for installing Paperless have [already been published](https://artifacthub.io/packages/search?ts_query_web=paperless&sort=relevance&page=1). Right now, we don't make any mention of these in our installation documentation, or allude to Kubernetes at all.

Up until recently, most potential Kubernetes users might have easily found [the popular Paperless Helm chart provided by the k8s-at-home project](https://github.com/k8s-at-home/charts/tree/master/charts/stable/paperless), which maintained a wide library of community Helm charts for commonly self-hosted services. Unfortunately, the k8s-at-home project [recently deprecated their charts repository](https://github.com/k8s-at-home/charts/issues/1761), and encouraged maintainers of the targeted software to take over ownership of the charts that k8s-at-home previously maintained.

I think we should fork the [k8s-at-home Paperless chart](https://github.com/k8s-at-home/charts/tree/master/charts/stable/paperless), officially maintain it, and document its usage as part of our installation guides.

### Licensing

Charts provided in the k8s-at-home library are [Apache 2.0 licensed](https://github.com/k8s-at-home/charts/blob/master/LICENSE). Paperless is [GPLv3 software](https://github.com/paperless-ngx/paperless-ngx/blob/main/LICENSE). The Apache Software Foundation has published [this document](https://www.apache.org/licenses/GPL-compatibility.html) on the compatibility between the licenses. The headline I extract from it is this: Apache 2.0 licensed software can be *included* in, but *not modified* in, GPLv3 licensed work.

The k8s-at-home project has encouraged other maintainers to take up their charts. It's possible they could provide release, to permit us to derive from the existing Paperless chart.

Regardless, I think a possible approach could be to fork the existing Paperless chart, and to retain its original license, with a statement to that effect included in each of the constituent files. This could complicate the licensing properties of this repository, but would allow us to keep the chart here alongside the software source, which is desirable.

I'm not a lawyer, nor have any particular background in this area, so I'd appreciate more eyes on this.

### Approach

Licensing hurdles aside, I think this is the rough approach we should take to implementation:

1. Copy the [k8s-at-home chart](https://github.com/k8s-at-home/charts/tree/master/charts/stable/paperless) and sources to this repository[1]. Change the name (to `paperless-ngx`) and metadata. Increment the version number[2].
2. Use the [helm chart-releaser-action](https://github.com/helm/chart-releaser-action) with GitHub Actions CI to automatically publish the Helm chart to GitHub Pages. While we don't use Pages (to my knowledge), this should be compatible with any future work we do there. Ultimately, this generates Helm package archives and a static index for Helm clients to use.
3. Add CI testing to ensure that the Helm chart behaves as expected. The downstream chart already has configuration for [chart-testing](https://github.com/helm/chart-testing), which has a corresponding GitHub action. I'm not sure exactly how this would look, but definitely something we should examine before taking the final step.
4. Once our new Helm chart repository is usable, add a documentation section to our install guide alongside the Docker instructions containing details on how to add the new Helm chart, and where to look for options. (Helm chart docs are often kept alongside the chart itself, and rendered in the repository browser. I believe this takes advantage of docs-generation tools, but I'm not certain.)

[1]: Using the same repository for the chart and source code has a number of benefits, just like keeping our Dockerfile here. [This article](https://insights.project-a.com/whats-the-best-way-to-manage-helm-charts-1cbf2614ec40) covers it well.
[2]: Existing installations should be able to change the source for their chart and upgrade smoothly.

### This Pull Request

In the first commit of this WIP pull request, I'm addressing points 1 and 2 above in the Approach section. I'm leaving point 3 until after we are able to discuss the licensing implications, and confirm that it behaves.

Fixes #1790 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by alexander-bauer** [comment]
Looks like my pre-commit checks failed to run in an odd way the first time. The affected `README.md` is automatically generated by `helm-docs`, so I think it should be ignored by `prettier`.

**Comment by shamoon** [comment]
I know basically nothing about k8s, can I just ask what are the pros/cons of putting this in a separate repo where it can be maintained on its own?

**Comment by alexander-bauer** [comment]
> I know basically nothing about k8s, can I just ask what are the pros/cons of putting this in a separate repo where it can be maintained on its own?

Totally fair question. Helm charts bear a lot of resemblance to Dockerfiles: while they're really a packaging artifact, there are some substantial benefits to keeping them in the main repository. These are, in broad strokes (**pros**):
- Continuous deployment (Chart version marches forward with application version)
- Application changes which require deployment methodology changes (like adding Redis as an external caching layer would have been) can alter the distribution packages, too
- Centralized issue management (Folks who have problems, whether with the Helm chart or the application itself, both arrive at this repository to report their issue; this is much nicer from a user's perspective)
- No need to manage a separate repository; this is a small concern as a group that already has a GitHub organization, but still reduces the permission-management complexity

As far as drawbacks, there's just one big one (**cons**):
- Licensing.

  The Helm chart from k8s-at-home is licensed under Apache 2.0. While k8s-at-home is decidedly amenable to projects in our position taking over the charts they've previously stewarded, I think the chart licensing is there to stay. If we keep it as part of this repository, we'll have to be clear about its separate (albeit more permissive) licensing.

Overall, I think that keeping it in a separate repository could lead to it rotting, without the visibility (and automation-assisted management tools) that it would have here. Just like with Dockerfiles, I really wouldn't even consider the possibility of keeping it separate if it weren't for the licensing issue; but I don't think that's an insurmountable hurdle.

**Comment by stumpylog** [comment]
I agree the licenses seem compatible, as long as it's kept clear as possible.

Looking at the implementation side a little more:

- To ignore the generated readme, change the `exclude` for prettier to `exclude: "(^Pipfile\\.lock$)|(charts/paperless-ngx/README.md)"` or the like
- I notice `helm/chart-releaser-action` creates a release each time, could that be confusing to have these releases interleaved?  It may be a reason to keep it separated?
- Maybe the trigger should be for tags instead of main, as tags are actual releases or release candidates?  This would reduce the interleaving of the above

**Comment by alexander-bauer** [comment]
All good points.

On the separation of the Ansible materials, I have a couple thoughts. (I wore the Ansible hat for a few years in my day job, so I've at least encountered this from the other side of the table.)

- More philosophically, I think there's a distinction, if blurry, between packaging code and "installer" code, for which Ansible falls into the latter category. Any schematic associated with "packaging" is usually used to render some single kind of archive, like `setup.py` -> `mypackage.zip` or `.egg`, or `myrpm.specf` -> `mypackage.rpm` or `Chart.yaml` -> `myhelmrelease.tar.gz`; whereas an "installer" or configuration management declarations module often lives separately because each release doesn't need to produce an `installer-x.y.z` style artifact.
- More practically: Ansible modules often behave better as their own repositories; some of the Ansible tooling for using them expects them to have a number of files at the root of the repository, that would be confusing if kept with the source code.

I don't think either way would be a hard and fast rule, but my opinion (held more strongly as mainly a user of the Helm tool), is that Helm chart definitions belong in the main repository with the code, as separate.

-------

As a separate thought, I wonder if it would make sense to prompt in our bug report templates for information on the installation environment, as, like, a checkbox: `Baremetal` / `Docker on Linux` / `Docker Compose` / `Container Service Platform (e.g. ECS)` / `Kubernetes`. (Unless we do already. It's been a while since I've filed a bug report.) As the userbase expands, odds are good that we'll see questions from folks in more and more kinds of environments. perhaps with various levels of insight about how those platforms might affect their deployments. If we have that info off the bat, some of the break-fix type troubleshooting might be headed off with common platform issues.

--------

I've left the license alone for now. I'm willing to take a crack at finding the necessary verbiage to write about the main code license versus the Helm chart license, or other artifacts that may someday come around, but I don't know anything more about licensing than anyone else might, and I've not been as intimately involved in the last few months of operations as y'all have.

Let me know what y'all think I should do with it.

**Comment by qcasey** [comment]
Thank you for your perspective on the Ansible repo, that's helpful. 

> I wonder if it would make sense to prompt in our bug report templates for information on the installation environment, as, like, a checkbox

We do have that, see the [bug-report](https://github.com/paperless-ngx/paperless-ngx/blob/main/.github/ISSUE_TEMPLATE/bug-report.yml) template. Although it could benefit from a few more entries.

> Let me know what y'all think I should do with it.

I have no other issues with the licensing as this PR stands. Is everyone else happy with bringing this over?

**Comment by shamoon** [comment]
Yea I have no beef with it but it’s not really my end of the business.

Speaking of which, my biggest (not that it’s humongous) concern is maintenance, though hopefully sounds like it won’t need to much. To that end, @eingemaischt let us know if you would be interested in joining the org.

I do think there’s still some work left on the PR itself though, no?

**Comment by alexander-bauer** [comment]
@shamoon Other than the licensing, I believe this is in a state to go live -- it looks like the CI tests are failing for unrelated reasons, or at least `Build Documentation` is. The checklist on the original PR is missing the "make changes to documentation" field, which I'm not sure needs to be part of this PR. As it stands, I don't believe there's any reference to the Helm chart in the documentation, so pulling this in wouldn't make anything wrong. (There still needs to be a change to write that documentation in the first place, but I think that's a separate addition to be made.)

Maintenance-wise, there shouldn't be much to do. The chart is a quite-simple Kubernetes app ("run these containers and connect them this way"), which shouldn't change much without drastic architectural alterations to the core application. If we did something like drop Redis altogether, then I think the chart would need tweaking, but that's something that us Kubernetes users can pretty easily contribute.

**Comment by shamoon** [comment]
Oh ok I didn't realize, yes agree not all those checkboxes apply :)

I'll take a look later!

**Review by shamoon — DISMISSED** [review]
Appreciate your patience with the PR.

**Comment by shamoon** [comment]
Tests passing OK after rebase. We recently moved our docs to gh-pages, I think this should be OK with that though? Looks like it will create 'index.yaml'. Ps. The docs for that action https://helm.sh/docs/howto/chart_releaser_action/ state "Note: The Chart Releaser Action is almost always used in tandem with the [Helm Testing Action](https://github.com/marketplace/actions/helm-chart-testing) and [Kind Action](https://github.com/marketplace/actions/kind-cluster)."

**Comment by alexander-bauer** [comment]
@shamoon I've used the action with gh-pages before - it works beautifully. All it should do is create that `index.yaml` and the packaged Helm chart `.tar.gz`, and commit them on that branch. I'd love to add testing at some point with those actions, too.

**Comment by shamoon** [comment]
Cool thanks 😊

I’ll leave it with the ci squad for final 👌

File: .github/workflows/release-chart.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    tags:
+      - v*
+
+jobs:
+  release_chart:
+    name: "Release Chart"
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
```
Comment (stumpylog) [inline .github/workflows/release-chart.yml:15]: Super minor, but this is a little behind, there's a v3 now

**Review by stumpylog — DISMISSED** [review]
Just a super minor thing that can be fixed or some other time.  Seems alright to me

File: .github/workflows/release-chart.yml (hunk 1 above)
Comment (shamoon) [inline .github/workflows/release-chart.yml:15]:
```suggestion
        uses: actions/checkout@v3
```

**Comment by stevijo** [comment]
Hey all,
just happened that i wanted to test this project out today and deploy it on my k8s cluster.

I really like the idea of having your own chart and thanks for adding this. The problem with this one is, that you are using the deprecated/archived library chart of k8s-at-home.
Not only https://github.com/k8s-at-home/charts is deprecated, but also https://github.com/k8s-at-home/library-charts moved to a new home: https://github.com/bjw-s/helm-charts/tree/main/charts/library/common. Is it possible to switch to the maintained version of the library chart - like recommended in the k8s-at-home issue https://github.com/k8s-at-home/charts/issues/1761?

**Comment by tooomm** [comment]
@alexander-bauer Since this is merged now, can you please check if it really works?

To me, it sounds like there should be an conflict with our new documentation. It utilizes GitHub's possibility to host static webpages (GitHub Pages) already as mentioned here. Not so sure if your changes works as you need them to as there is never a GitHub Pages instance deployed for the helm chart: https://github.com/paperless-ngx/paperless-ngx/deployments
The reason is that the repo is configured to deploy our own documentation, and not the default location where your action pushes content to (gh-pages branch).

**Comment by stumpylog** [comment]
It doesn't appear the action took any action: https://github.com/paperless-ngx/paperless-ngx/actions/runs/3615048321/jobs/6091895071

Maybe related: helm/chart-releaser-action/issues/120

**Comment by tooomm** [comment]
That's right stumpylog! 🙈😄 
I got confused by the content in gh-pages branch and did not realize that's where our own docs are deployed from.
If this only adds a yml file there it might not be an issue.

Reading at https://helm.sh/docs/howto/chart_releaser_action/, it says that the action will create releases as well. 🤔

**Comment by alexander-bauer** [comment]
@stevijo Good catch, thank you! I'll adjust it to use the correct upstream chart.

@stumpylog @tooomm I think that the action is set to only run on tags - I would guess that it'll publish the chart as soon as the next release is tagged. I'm not sure what the best way to make that happen is, or whether there's a way to run the action manually.

**Comment by stumpylog** [comment]
That linked run was from a tag push.  From the logs, it didn't think anything changed, but I'm not sure what it's looking for. It may be a conflict with what the Git history looks like on the gh-pages branch.

**Comment by shamoon** [comment]
Isnt the issue just that there was no pre-existing `index.yaml`?

**Comment by alexander-bauer** [comment]
I think that the Chart releaser action compares the chart version (from `Chart.yaml`) to that from the previous commit, and only takes action if they differ. I think it only looks at the head and previous commit, but this PR included multiple commits with the chart unchanged. I'm about to submit a new PR that just bumps the minor chart version with no other change; that should hopefully trigger it.

**Pull request merged**